bundles/org.openhab.binding.argoclima/src/main/java/org/openhab/binding/argoclima/internal/configuration/ArgoClimaConfigurationBase.java

```diff
@@ -0,0 +1,374 @@
+                getExtraFieldDescription(), deviceCpuId, refreshInterval, oemServerPort,
+                getOrDefault(this::getOemServerAddress),
+                getOrDefault(() -> getScheduleDayOfWeek(ScheduleTimerType.SCHEDULE_1)),
+                getOrDefault(() -> getScheduleOnTime(ScheduleTimerType.SCHEDULE_1)),
+                getOrDefault(() -> getScheduleOffTime(ScheduleTimerType.SCHEDULE_1)),
+                getOrDefault(() -> getScheduleDayOfWeek(ScheduleTimerType.SCHEDULE_2)),
+                getOrDefault(() -> getScheduleOnTime(ScheduleTimerType.SCHEDULE_2)),
+                getOrDefault(() -> getScheduleOffTime(ScheduleTimerType.SCHEDULE_2)),
+                getOrDefault(() -> getScheduleDayOfWeek(ScheduleTimerType.SCHEDULE_3)),
+                getOrDefault(() -> getScheduleOnTime(ScheduleTimerType.SCHEDULE_3)),
+                getOrDefault(() -> getScheduleOffTime(ScheduleTimerType.SCHEDULE_3)), resetToFactoryDefaults);
+    }
+
+    /**
+     * Return derived class'es extra configuration parameters (for a common {@link toString} implementation)
+     *
+     * @return Comma-separated list of configuration parameter=value pairs or empty String if derived class does not
+     *         introduce any.
+     */
+    protected abstract String getExtraFieldDescription();
+
+    /**
+     * Validate derived configuration
+     *
+     * @throws ArgoConfigurationException - on validation failure
+     */
+    protected abstract void validateInternal() throws ArgoConfigurationException;
+
+    /**
+     * Validate current config
+     *
+     * @return Error message if config is invalid. Empty string - otherwise
+     */
+    public final String validate() {
+        if (refreshInterval < 0) {
+            return "Refresh interval must be >= 0";
+        }
+
+        if (oemServerPort < 0 || oemServerPort >= 65536) {
+            return "OEM server port must be in range [0..65536]";
+        }
+
+        try {
+            // want the side-effect of these calls
+            getOemServerAddress();
+
+            getScheduleDayOfWeek(ScheduleTimerType.SCHEDULE_1);
+            getScheduleOnTime(ScheduleTimerType.SCHEDULE_1);
+            getScheduleOffTime(ScheduleTimerType.SCHEDULE_1);
+
+            getScheduleDayOfWeek(ScheduleTimerType.SCHEDULE_2);
+            getScheduleOnTime(ScheduleTimerType.SCHEDULE_2);
+            getScheduleOffTime(ScheduleTimerType.SCHEDULE_2);
+
+            getScheduleDayOfWeek(ScheduleTimerType.SCHEDULE_3);
+            getScheduleOnTime(ScheduleTimerType.SCHEDULE_3);
+            getScheduleOffTime(ScheduleTimerType.SCHEDULE_3);
+
+            validateInternal();
+            return "";
+        } catch (Exception e) {
+            var msg = Optional.ofNullable(e.getLocalizedMessage());
+            var cause = Optional.ofNullable(e.getCause());
+            return msg.orElse("Null exception message")
```

Comment:
Is it only possible to have a null exception message? If so, it would be better to check/handle the null exception up front.

Reply:
It's just boilerplate, really. I'm not throwing any exceptions with `null` messages (in fact, all the custom exceptions have `@NonNull` annotation there), and all that could happen here are typical framework exceptions like parse errors. None of these have null messages **in practice**.

However, this is a generic catch, theoretically the interface allows `null`s and I want to:
- keep SpotBugs happy
- keep this method as a strict **no-throw**

So, this message is just boilerplate I don't **expect** to be hit **ever** (unless I have a bug), which is also why I didn't bother to localize this message. 
I'll add the same explanation to comments to keep it explicit, but am inclined to keep the code as-is.
If you have a specific refactor in mind, please suggest.